File: src/wiki_builder.py
"""Génération du wiki type portail d'actualités."""

from __future__ import annotations

import re
from datetime import datetime, timezone

from .fetcher import Article
from .gemini_briefing import BriefingResult, ProcessedArticle, rubrique_slug


def _format_date(dt) -> str:
    if not dt:
        return ""
    return dt.strftime("%d/%m %H:%M")


def _article_block(p: ProcessedArticle) -> list[str]:
    a = p.article
    lines = [
        f"### {a.title}",
        "",
        f"**{a.feed_name}** · {_format_date(a.published)} · [Lire]({a.link})",
        "",
    ]
    if a.image_url:
        lines.append(f"![{a.title}]({a.image_url})")
        lines.append("")
    lines.append(p.summary)
    lines.append("")
    lines.append("---")
    lines.append("")
    return lines


def render_home_portal(briefing: BriefingResult) -> str:
    """Page d'accueil wiki — portail d'actualités."""
    now = datetime.now(timezone.utc)
    total = len(briefing.all_articles)
    rubriques = sorted(briefing.by_rubrique.keys(), key=lambda r: -len(briefing.by_rubrique[r]))

    lines = [
        f"# 📰 Actus — {briefing.date_str}",
        "",
        f"_Briefing personnel · {total} articles · mis à jour le "
        f"{now.strftime('%d/%m/%Y à %H:%M')} (UTC)_",
        "",
        "> Page d'accueil : choisissez une rubrique ou parcourez les titres ci-dessous.",
        "",
        "## 🔥 À la une",
        "",
    ]

    for p in briefing.all_articles[:6]:
        a = p.article
        lines.append(f"- **[[{rubrique_slug(p.rubrique)}|{p.rubrique}]]** — "
                     f"{a.title} — [lire]({a.link})")

    lines.extend(["", "## 📂 Toutes les rubriques", "", "| Rubrique | Articles |", "|----------|----------|"])
    for rub in rubriques:
        slug = rubrique_slug(rub)
        count = len(briefing.by_rubrique[rub])
        lines.append(f"| [[{slug}|{rub}]] | {count} |")

    lines.extend(["", "## 📋 Derniers articles", ""])
    for p in briefing.all_articles[:15]:
        a = p.article
        lines.append(f"- **{a.title}** _({p.rubrique})_ — [article]({a.link})")

    lines.extend([
        "",
        "---",
        "",
        f"Archive du jour : [[Briefing-{briefing.date_str}]]",
        "",
        "_Généré par Actus Sync (Android) · Le Monde · Charente Libre · Gemini_",
    ])
    return "\n".join(lines) + "\n"


def render_rubrique_page(rubrique: str, items: list[ProcessedArticle]) -> str:
    lines = [
        f"# {rubrique}",
        "",
        f"_{len(items)} article(s) — retour [[Home]]_",
        "",
    ]
    for p in items:
        lines.extend(_article_block(p))
    return "\n".join(lines).rstrip() + "\n"


def render_full_briefing(briefing: BriefingResult) -> str:
  lines = [
      f"# Briefing complet — {briefing.date_str}",
      "",
      f"_{len(briefing.all_articles)} articles classés automatiquement_",
      "",
  ]
  for rubrique in sorted(briefing.by_rubrique.keys()):
      lines.append(f"## {rubrique}")
      lines.append("")
      for p in briefing.by_rubrique[rubrique]:
          lines.extend(_article_block(p))
  return "\n".join(lines).rstrip() + "\n"


def build_wiki_files(briefing: BriefingResult) -> dict[str, str]:
    """Retourne {nom_fichier.md: contenu}."""
    files: dict[str, str] = {"Home.md": render_home_portal(briefing)}
    for rubrique, items in briefing.by_rubrique.items():
        if not items:
            continue
        slug = rubrique_slug(rubrique)
        # Nom de page wiki lisible
        page_name = slug.replace("-", " ").title().replace(" ", "-")
        if rubrique == "Idées / Débats":
            page_name = "Idees-Debats"
        elif rubrique == "Faits divers":
            page_name = "Faits-divers"
        files[f"{page_name}.md"] = render_rubrique_page(rubrique, items)

    files[f"Briefing-{briefing.date_str}.md"] = render_full_briefing(briefing)
    return files

File: android/app/src/main/kotlin/fr/actus/sync/data/WikiBuilder.kt
package fr.actus.sync.data

import java.time.Instant
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter

object WikiBuilder {
    private val dateTimeFmt = DateTimeFormatter.ofPattern("dd/MM HH:mm").withZone(ZoneOffset.UTC)
    private val fullDateFmt = DateTimeFormatter.ofPattern("dd/MM/yyyy 'à' HH:mm").withZone(ZoneOffset.UTC)

    fun buildWikiFiles(briefing: BriefingResult): Map<String, String> {
        val files = linkedMapOf("Home.md" to renderHomePortal(briefing))
        briefing.byRubrique.forEach { (rubrique, items) ->
            if (items.isEmpty()) return@forEach
            files[pageFileName(rubrique)] = renderRubriquePage(rubrique, items)
        }
        files["Briefing-${briefing.dateStr}.md"] = renderFullBriefing(briefing)
        return files
    }

    fun rubriqueSlug(name: String): String =
        name.lowercase()
            .replace(Regex("[^\\w\\s-]"), "")
            .replace(Regex("[-\\s]+"), "-")
            .trim('-')
            .ifBlank { "autres" }

    private fun pageFileName(rubrique: String): String = when (rubrique) {
        "Idées / Débats" -> "Idees-Debats.md"
        "Faits divers" -> "Faits-divers.md"
        else -> rubriqueSlug(rubrique)
            .split('-')
            .joinToString("-") { part -> part.replaceFirstChar { c -> c.titlecase() } } + ".md"
    }

    private fun formatDate(instant: Instant?): String =
        if (instant == null) "" else dateTimeFmt.format(instant)

    private fun articleBlock(processed: ProcessedArticle): String {
        val article = processed.article
        val lines = buildList {
            add("### ${article.title}")
            add("")
            add("**${article.feedName}** · ${formatDate(article.published)} · [Lire](${article.link})")
            add("")
            if (article.imageUrl.isNotBlank()) {
                add("![${article.title}](${article.imageUrl})")
                add("")
            }
            add(processed.summary)
            add("")
            add("---")
            add("")
        }
        return lines.joinToString("\n")
    }

    private fun renderHomePortal(briefing: BriefingResult): String {
        val now = Instant.now()
        val total = briefing.allArticles.size
        val rubriques = briefing.byRubrique.keys.sortedByDescending { briefing.byRubrique[it]?.size ?: 0 }

        val lines = buildList {
            add("# 📰 Actus — ${briefing.dateStr}")
            add("")
            add(
                "_Briefing personnel · $total articles · mis à jour le " +
                    "${fullDateFmt.format(now)} (UTC)_",
            )
            add("")
            add("> Page d'accueil : choisissez une rubrique ou parcourez les titres ci-dessous.")
            add("")
            add("## 🔥 À la une")
            add("")
            briefing.allArticles.take(6).forEach { processed ->
                val article = processed.article
                add(
                    "- **[[${rubriqueSlug(processed.rubrique)}|${processed.rubrique}]]** — " +
                        "${article.title} — [lire](${article.link})",
                )
            }
            add("")
            add("## 📂 Toutes les rubriques")
            add("")
            add("| Rubrique | Articles |")
            add("|----------|----------|")
            rubriques.forEach { rub ->
                val count = briefing.byRubrique[rub]?.size ?: 0
                add("| [[${rubriqueSlug(rub)}|$rub]] | $count |")
            }
            add("")
            add("## 📋 Derniers articles")
            add("")
            briefing.allArticles.take(15).forEach { processed ->
                val article = processed.article
                add("- **${article.title}** _(${processed.rubrique})_ — [article](${article.link})")
            }
            add("")
            add("---")
            add("")
            add("Archive du jour : [[Briefing-${briefing.dateStr}]]")
            add("")
            add("_Généré par Actus Sync (Android) · Le Monde · Charente Libre · Gemini_")
        }
        return lines.joinToString("\n") + "\n"
    }

    private fun renderRubriquePage(rubrique: String, items: List<ProcessedArticle>): String {
        val lines = buildList {
            add("# $rubrique")
            add("")
            add("_${items.size} article(s) — retour [[Home]]_")
            add("")
            items.forEach { add(articleBlock(it)) }
        }
        return lines.joinToString("\n").trimEnd() + "\n"
    }

    private fun renderFullBriefing(briefing: BriefingResult): String {
        val lines = buildList {
            add("# Briefing complet — ${briefing.dateStr}")
            add("")
            add("_${briefing.allArticles.size} articles classés automatiquement_")
            add("")
            briefing.byRubrique.keys.sorted().forEach { rubrique ->
                add("## $rubrique")
                add("")
                briefing.byRubrique[rubrique].orEmpty().forEach { add(articleBlock(it)) }
            }
        }
        return lines.joinToString("\n").trimEnd() + "\n"
    }
}
